tests/test_reduction_ops.py

```diff
@@ -689,6 +689,36 @@ def test_accuracy_index_select(shape, dim, dtype):
     gems_assert_equal(res_out, ref_out)
 
 
+@pytest.mark.index_select_backward
+@pytest.mark.parametrize("shape", REDUCTION_SHAPES)
+@pytest.mark.parametrize("dim", DIM_LIST)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_index_select_backward(shape, dim, dtype):
+    inp = torch.randn(shape, dtype=dtype, device="cuda", requires_grad=True)
+    ref_inp = to_reference(inp)
+    from math import floor
+
+    index_size = inp.size(dim)
+    index = torch.randint(0, index_size, [floor(index_size * 0.8)], device="cuda")
+    index = torch.unique(index)
```

Comment:
What if there is duplicated index? I don't think current implementation can handle this.

Reply:
Yes, for duplicted index , torch.autograd.grad can handle this case, our design code also can handle. But for accuracy test,  out_grad and index random generate ,not meet same index and value keep correspond's condition. So, we add the unique statement.；

 index
![1](https://github.com/user-attachments/assets/45206cc8-076f-4bb1-97d5-fa7cc0e99b83)

out_grad
![2](https://github.com/user-attachments/assets/8dc86b61-a312-43dc-ab98-87d6413fc342)

normally ,the index is duplicated,the correwponding value overwritten, and the result is correct.

Reply:
No, I don't see how it can handle duplicated indices. With duplicated indices, a sub-tensor is selected multiple times in the forward op, thus the corresponding gradient should accumulate into the corresponding input gradient.

There is no such accumulation in the kernel. I add a test, and it always fails.

```python
@pytest.mark.index_select_backward
@pytest.mark.parametrize("shape", REDUCTION_SHAPES)
@pytest.mark.parametrize("dim", DIM_LIST)
@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
def test_accuracy_index_select_backward(shape, dim, dtype):
    inp = torch.randn(shape, dtype=dtype, device="cuda", requires_grad=True)
    ref_inp = to_reference(inp)
    from math import floor

    index_size = inp.size(dim)
    index = torch.tensor([0, 0, 0, 0], device="cuda")
    # index = torch.unique(index)
    if len(index) == 0:
        pass
    else:
        ref_index = to_reference(index)
        ref_out = torch.index_select(ref_inp, dim, ref_index)
        out_grad = torch.randn_like(ref_out)
        ref_grad = to_reference(out_grad)
        (ref_in_grad,) = torch.autograd.grad(ref_out, ref_inp, ref_grad)

        with flag_gems.use_gems():
            res_out = torch.index_select(inp, dim, index)
            (res_in_grad,) = torch.autograd.grad(res_out, inp, out_grad)
        res_out = to_reference(res_out)
        res_in_grad = to_reference(res_in_grad)
        gems_assert_equal(res_out, ref_out)
        gems_assert_equal(res_in_grad, ref_in_grad)
```

Reply:
out_grad !=ref_out

Reply:
ok